src/graph/query_graph.c

```diff
@@ -93,6 +102,99 @@ static void _BuildQueryGraphAddEdge(QueryGraph *qg, const cypher_astnode_t *ast_
 	else QueryGraph_ConnectNodes(qg, dest, src, edge);
 }
 
+// Extracts node from 'qg' and places a copy of into 'graph'
+static void _QueryGraph_ExtractNode(const QueryGraph *qg, QueryGraph *graph,
+		AST *ast, const cypher_astnode_t *ast_node) {
+
+	// Validate inputs.
+	ASSERT(qg != NULL && graph != NULL && ast != NULL && ast_node != NULL);
+
+	// See if node is already in 'graph'.
+	const char *alias = AST_GetEntityName(ast, ast_node);
+	QGNode *n = QueryGraph_GetNodeByAlias(graph, alias);
+
+	if(n == NULL) {
+		// Node is missing from 'graph', try getting it from 'qg'.
+		n = QueryGraph_GetNodeByAlias(qg, alias);
+		if(n == NULL) {
```

Comment:
The conditions for triggering this are a bit confusing. So far as I can tell, in a query like:
`MATCH (a) WITH a WHERE (a)-[]->(:L1) RETURN a`
The QueryGraph for the first scope only contains `a`, while the QueryGraph for the second scope contains `a` and the anonymous edge and node. In filter op placement (and subsequently `ExecutionPlan_ReduceFilterToApply`), we move the WHERE logic into the first scope, and as such have to extend the QueryGraph.

I've managed to trigger one assertion failure based on this oddness, see comment below.

In any case, I think this logic should have more explanatory comments for why we might be here.

Reply:
👍